File: barkr/models/mention_style.py
"""
Module to implement an enum representing how mentions should be
rendered when serializing a message's body for a given publisher.
"""

from enum import Enum
from typing import Optional

from barkr.models.message_mention import MessageMention


class MentionStyle(Enum):
    """
    Represents how mentions should be rendered when serializing
    a message's body for a given publisher.
    """

    # Leave the message body untouched
    PLAIN = 0

    # Keep the original handle text and append the profile URL in parentheses
    APPEND_URL = 1

    # Replace the original handle text with the profile URL
    REPLACE_WITH_URL = 2

    # Render mentions as markdown links with the username as link text
    # and the profile URL as the link target
    MARKDOWN_LINK = 3

    # Render mentions as HTML links with the username as link text and the
    # profile URL as the href target
    HTML_LINK = 4

    @staticmethod
    def replace_mentions(
        text: str,
        mentions: Optional[list[MessageMention]],
        mention_style: Optional["MentionStyle"] = None,
    ) -> str:
        """
        Replace mentions in a text body according to a render style.

        :param text: The source text to rewrite
        :param mentions: The mention metadata extracted from the source post
        :param mention_style: The style to use when rendering mentions
        :return: The rewritten text
        """

        if mention_style is None:
            mention_style = MentionStyle.PLAIN

        if mention_style == MentionStyle.PLAIN or not mentions:
            return text

        content_parts: list[str] = []
        cursor = 0
        for mention in mentions:
            start = text.find(mention.username, cursor)
            if start == -1:
                continue

            content_parts.append(text[cursor:start])
            if mention_style == MentionStyle.APPEND_URL:
                replacement = f"{mention.username} ({mention.url})"
            elif mention_style == MentionStyle.REPLACE_WITH_URL:
                replacement = mention.url
            elif mention_style == MentionStyle.MARKDOWN_LINK:
                replacement = f"[{mention.username}]({mention.url})"
            elif mention_style == MentionStyle.HTML_LINK:
                replacement = f'<a href="{mention.url}">{mention.username}</a>'
            else:
                raise ValueError(f"Unsupported mention style: {mention_style}")

            content_parts.append(replacement)
            cursor = start + len(mention.username)

        content_parts.append(text[cursor:])
        return "".join(content_parts)

File: models/mention_style.go
package models

import (
	"fmt"
	"strings"
)

// MentionStyle represents how user mentions should be rendered when
// serializing a [Message] for a given [Connection]. Each [Connection] implementation
// can decide how to display mentions from an incoming [Message] based on what
// they support.
type MentionStyle int

const (
	// MentionStylePlain indicates that the [Message] is rendered as-is, with no mention enrichment
	MentionStylePlain MentionStyle = iota
	// MentionStyleAppendURL indicates that user mentions on the [Message] are enriched with the profile URL in parenthesis next to it
	MentionStyleAppendURL
	// MentionStyleReplaceWithURL indicates that user mentions on the [Message] are replaced with the profile URL
	MentionStyleReplaceWithURL
	// MentionStyleMarkdownLink indicates that user mentions on the [Message] are linked to the profile URL using Markdown syntax
	MentionStyleMarkdownLink
	// MentionStyleHTMLLink indicates that user mentions on the [Message] are linked to the profile URL using HTML syntax (anchor tag)
	MentionStyleHTMLLink
)

// String returns the string representation of the [MentionStyle].
func (m MentionStyle) String() string {
	switch m {
	case MentionStylePlain:
		return "plain"
	case MentionStyleAppendURL:
		return "append-url"
	case MentionStyleReplaceWithURL:
		return "replace-with-url"
	case MentionStyleMarkdownLink:
		return "markdown-link"
	case MentionStyleHTMLLink:
		return "html-link"
	default:
		return fmt.Sprintf("unknown (%d)", m)
	}
}

// ReplaceMentions replaces or enriches user mentions in a given text according
// to the specified [MentionStyle].
func ReplaceMentions(text string, mentions []MessageMention, style MentionStyle) (string, error) {
	if style == MentionStylePlain || len(mentions) == 0 {
		return text, nil
	}

	var contentParts []string

	cursor := 0
	for _, mention := range mentions {
		index := strings.Index(text[cursor:], mention.Username)

		if index == -1 {
			continue
		}

		start := cursor + index
		contentParts = append(contentParts, text[cursor:start])

		var replacement string
		switch style {
		case MentionStyleAppendURL:
			replacement = fmt.Sprintf("%s (%s)", mention.Username, mention.URL)
		case MentionStyleReplaceWithURL:
			replacement = mention.URL
		case MentionStyleMarkdownLink:
			replacement = fmt.Sprintf("[%s](%s)", mention.Username, mention.URL)
		case MentionStyleHTMLLink:
			replacement = fmt.Sprintf("<a href=\"%s\">%s</a>", mention.URL, mention.Username)
		default:
			return "", fmt.Errorf("unsupported mention style: %s", style)
		}

		contentParts = append(contentParts, replacement)
		cursor = start + len(mention.Username)
	}

	contentParts = append(contentParts, text[cursor:])
	return strings.Join(contentParts, ""), nil
}
